src/zeta/aca_oam_server.cpp

```diff
@@ -0,0 +1,323 @@
+    return;
+  }
+
+  oammsg = (oam_message *)message;
+
+  if (_validate_oam_message(oammsg)) {
+    ACA_LOG_ERROR("%s", "Invalid OAM message!\n");
+    return;
+  }
+
+  uint8_t msg_type = (uint8_t)_get_message_type(oammsg);
+
+  (this->*_parse_oam_msg_ops[msg_type])(udp_dport, oammsg);
+
+  return;
+}
+
+void ACA_Oam_Server::_init_oam_msg_ops()
+{
+  _parse_oam_msg_ops[OAM_MSG_FLOW_INJECTION] =
+          &aca_oam_server::ACA_Oam_Server::_parse_oam_flow_injection;
+  _parse_oam_msg_ops[OAM_MSG_FLOW_DELETION] =
+          &aca_oam_server::ACA_Oam_Server::_parse_oam_flow_deletion;
+  _parse_oam_msg_ops[OAM_MSG_NONE] = &aca_oam_server::ACA_Oam_Server::_parse_oam_none;
+}
+
+uint8_t ACA_Oam_Server::_get_message_type(oam_message *oammsg)
+{
+  if (!oammsg) {
+    ACA_LOG_ERROR("%s", "OAM message is null!\n");
+    return OAM_MSG_NONE;
+  }
+
+  if (!oammsg->op_code) {
+    return OAM_MSG_NONE;
+  }
+
+  return (uint8_t)(ntohl(oammsg->op_code));
+}
+
+string ACA_Oam_Server::_get_mac_addr(uint8_t *mac)
+{
+  string mac_string;
+  stringstream ss;
+
+  //Convert mac address to string
+  // from uint8[6] to string
+  for (int i = 0; i < 6; i++) {
+    ss << std::hex << std::setw(2) << std::setfill('0')
+       << static_cast<unsigned int>(mac[i]);
+    ss << ":";
+  }
+  ss >> mac_string;
+  mac_string.pop_back();
+
+  return mac_string;
+}
+
+void ACA_Oam_Server::_standardize_mac_address(string &mac_string)
+{
+  // standardize the mac address to aa:bb:cc:dd:ee:ff
+  std::transform(mac_string.begin(), mac_string.end(), mac_string.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+  std::replace(mac_string.begin(), mac_string.end(), '-', ':');
+}
+
+string ACA_Oam_Server::_get_vpc_id(uint8_t *vni)
+{
+  string vpc_id;
+  stringstream ss;
+
+  for (int i = 0; i < 3; i++) {
+    ss << std::hex << std::setw(2) << std::setfill('0')
+       << static_cast<unsigned int>(vni[i]);
+  }
+
+  ss >> vpc_id;
+  vpc_id.pop_back();
+
+  return vpc_id;
+}
+
+//extract data for flow table matching from the oam message
+oam_match ACA_Oam_Server::_get_oam_match_field(oam_message *oammsg)
+{
+  oam_match match;
+
+  flow_inject_msg msg_data = oammsg->data.msg_inject_flow;
+
+  match.sip = inet_ntoa(msg_data.inner_src_ip);
+  match.dip = inet_ntoa(msg_data.inner_dst_ip);
+  match.sport = to_string(ntohs(msg_data.src_port));
+  match.dport = to_string(ntohs(msg_data.dst_port));
+  match.proto = to_string(msg_data.proto);
+  match.vni = _get_vpc_id(msg_data.vni);
+
+  return match;
+}
+
+//extract the data used for the flow table action from the oam message
+oam_action ACA_Oam_Server::_get_oam_action_field(oam_message *oammsg)
+{
+  oam_action action;
+
+  flow_inject_msg msg_data = oammsg->data.msg_inject_flow;
+
+  action.inst_nw_dst = inet_ntoa(msg_data.inst_dst_ip);
+  action.node_nw_dst = inet_ntoa(msg_data.node_dst_ip);
+  action.inst_dl_dst = _get_mac_addr(msg_data.inst_dst_mac);
+  _standardize_mac_address(action.inst_dl_dst);
```

Comment:
@liangbin-pub - do we need to standardize the mac address received in oam? will Zeta always use the mac address format we prefer aa:bb:cc:dd:ee:ff

Reply:
per discussion, let's try to go without it.